src/cmd/flash.js

```diff
@@ -125,40 +126,82 @@ module.exports = class FlashCommand extends CLICommandBase {
 		const { api, auth } = this._particleApi();
 		const device = await usbUtils.getOneUsbDevice({ idOrName: deviceIdOrName, api, auth, ui: this.ui });
 
-		const platformName = platformForId(device.platformId).name;
+		const platformId = device.platformId;
+		const platformName = platformForId(platformId).name;
+		const currentDeviceOsVersion = device.firmwareVersion;
 
 		this.ui.write(`Flashing ${platformName} ${deviceIdOrName || device.id}`);
+
 		validateDFUSupport({ device, ui: this.ui });
 
 		let { skipDeviceOSFlash, files: filesToFlash } = await this._prepareFilesToFlash({
 			knownApp,
 			parsedFiles,
-			platformId: device.platformId,
+			platformId,
 			platformName,
 			target
 		});
 
 		filesToFlash = await this._processBundle({ filesToFlash });
 
 		const fileModules = await parseModulesToFlash({ files: filesToFlash });
-		await this._validateModulesForPlatform({ modules: fileModules, platformId: device.platformId, platformName });
+
+		const eligibleModules = await this._filterAssetsOnDevice(device, fileModules);
+
+		await this._validateModulesForPlatform({ modules: eligibleModules, platformId, platformName });
 
 		const deviceOsBinaries = await this._getDeviceOsBinaries({
-			currentDeviceOsVersion: device.firmwareVersion,
+			currentDeviceOsVersion,
 			skipDeviceOSFlash,
 			target,
-			modules: fileModules,
-			platformId: device.platformId,
+			modules: eligibleModules,
+			platformId,
 			applicationOnly
 		});
 		const deviceOsModules = await parseModulesToFlash({ files: deviceOsBinaries });
-		let modulesToFlash = [...fileModules, ...deviceOsModules];
-		modulesToFlash = filterModulesToFlash({ modules: modulesToFlash, platformId: device.platformId });
+		let modulesToFlash = [...eligibleModules, ...deviceOsModules];
+		modulesToFlash = filterModulesToFlash({ modules: modulesToFlash, platformId });
+
+		const flashSteps = await createFlashSteps({
+			modules: modulesToFlash,
+			isInDfuMode: device.isInDfuMode,
+			platformId
+		});
 
-		const flashSteps = await createFlashSteps({ modules: modulesToFlash, isInDfuMode: device.isInDfuMode , platformId: device.platformId });
 		await flashFiles({ device, flashSteps, ui: this.ui });
 	}
 
+	async _filterAssetsOnDevice(device, modules) {
+		if (device.isInDfuMode) {
+			device = await usbUtils.reopenInNormalMode(device, { reset: true });
```

Comment:
Let's use a different approach because this makes flashing less reliable. If the application firmware is crashing and someone manually puts the device in DFU mode to reflash it and the CLI reopens it in normal mode, the device will crash before it can get the asset info. This will also call `getAssetInfo` even in case the application doesn't have assets.

I think we could create the flash steps as if we're flashing all the assets, but in `flashFiles` when it comes time to flash an asset, call `getAssetInfo` (cache the result in case there are other assets) and skip calling `device.updateFirmware` if the asset already exists.